#2124: MySQL task

manesioz opened the pull request:
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

WIP PR for #2124 

- [x] adds new tests (if appropriate)
- [ ] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
Adds new task that enables users to execute and fetch queries from a MySQL DB. 


## Why is this PR important?
Increase useability due to integration with popular DB.

lauralorenz commented:
@manesioz Just a heads up that master has a fix in it to some tests that were failing due to an upgrade in an underlying dependency, you might want to merge/rebase master in before you get too detailed on the test failures to avoid any CircleCI noise!

manesioz commented:
hey @lauralorenz I seem to be failing the circle-ci pipeline due to the following error: 
```sh
_______________ ERROR collecting tests/tasks/mysql/test_mysql.py _______________
ImportError while importing test module '/root/project/tests/tasks/mysql/test_mysql.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
src/prefect/tasks/mysql/__init__.py:7: in <module>
    from prefect.tasks.mysql.mysql import MySQLExecute, MySQLFetch
src/prefect/tasks/mysql/mysql.py:1: in <module>
    import pymysql.cursors
E   ModuleNotFoundError: No module named 'pymysql'
```
Despite including `mypysql` in the `extras` of the `setup.py` file, and the tests passing locally. 

Before I continue further testing I was wondering if you had any idea why the environment doesn't seem to be picking up the dependency? Sorry, I don't have much experience with circle-ci.

joshmeek commented:
@manesioz for the `tests/tasks/mysql/__init__.py` add a pytest `importorskip` similar to https://github.com/PrefectHQ/prefect/blob/master/tests/tasks/snowflake/__init__.py 😄

manesioz commented:
Ah of course, thanks @joshmeek, sorry for overlooking that!

manesioz commented:
Hey @lauralorenz @joshmeek I think I have the first iteration done, seems to work fine locally so if you want to review whenever you have time feel free :)

lauralorenz commented:
@manesioz Hi! FYI I did start kicking the tires on this, I haven't finished yet but here is my setup:

* start a docker container of `mysql:latest`
* run this basic flow:

```python
import prefect
from prefect.tasks.mysql import MySQLExecute, MySQLFetch
from prefect import Flow

flow = Flow('my flow')
settings = {
'host': 'localhost',
'port': 3307,
'db_name': 'mysql',
'password': 'secret',
'user': 'root'
}
settings.update({'query': "SELECT * FROM user"})
s = MySQLFetch(**settings)
settings.update({'query': "SHOW databases;"})
r = MySQLExecute(**settings)

flow.add_task(s)
flow.add_task(r)
```

I got a couple errors, starting off with missing `cryptography`, and then charset not being passed to MySQLFetch. So there are a few little bits missing that might benefit from actually running this use case.

In addition I think we might be able to come up with a few more test cases, and potentially some integration tests for this if we want to go that way. For task library stuff I think that adding in more integration tests against docker containers would really increase confidence in the task library pieces. The postgres inspiration task doesn't really have these, but we could start layering that in here. I think we could add in some more test dependencies for this, like something that integrates docker containers with pytest, to set the stage for all task library tests if we want.

lauralorenz commented on src/prefect/tasks/mysql/mysql.py:
```diff
@@ -0,0 +1,192 @@
+        user: str,
+        password: str,
+        host: str,
+        port: int = 5432,
```
FYI the default in these spots should be 3307, as that is the default mysql port

lauralorenz commented on src/prefect/tasks/mysql/mysql.py:
```diff
@@ -0,0 +1,204 @@
+        - host (str): database host address
+        - port (int, optional): port used to connect to MySQL database, defaults to 5432 if not provided
+        - query (str, optional): query to execute against database
+        - data (tuple, optional): values to use in query, must be specified using placeholder in query string
```
```suggestion
        - data (tuple, optional, UNUSED): values to use in query, must be specified using placeholder in query string
```
Right now the `data` parameter is not being used by the implementation here, which I think we should either
a) fix in this PR
b) fix in a later PR, but remove the kwargs from here
c) fix in a later PR, and annotate the kwargs from this PR that they are unused

This github suggestion goes for (c) because it's the easiest to make as a one-line suggestion, but any of the 3 work for me depending what you want to tackle!

manesioz replied:
Good catch, sorry about that. I am going to remove it for this PR as it is not necessary, and was used previously for templating strings (but Python has evolved better ways of templating, such as `.format()` and f-strings)

manesioz commented on src/prefect/tasks/mysql/mysql.py:
```diff
@@ -0,0 +1,204 @@
+                        conm.commit()
+
+            conn.close()
+            print("Execute Results: ", executed)
```
You're absolutely right, I used it for debugging and forgot to port it over to proper logging. Thanks!

manesioz commented:
Hey @lauralorenz , thanks for the review! Sorry for not being as responsive, things have been hectic lately. I hope I addressed your comments sufficiently, and regarding the integration testing I was wondering if I could be pointed towards some examples of what you mean? 
I don't have much experience with database integration tests in Python so I am not quite sure what the best practices are.

bratao commented:
@manesioz @lauralorenz Is there a chance of MySQL task becoming official?

lauralorenz commented:
Hi, yes @bratao!

In retrospect I did this kind of overly complicated but I made some changes including an edit I just made adding the changelog in a PR https://github.com/manesioz/prefect/pull/1 to your remote @manesioz. I did that to keep them in queue for comment while I was waiting for the changelog updates but now I realized that I can't merge that PR, though I can make the commit changes to the branch directly 😅 So if I don't hear otherwise shortly I can just cherry pick them in.

manesioz commented:
Thanks @lauralorenz ! I appreciate your guidance and help with this.
